Language_Parser/actionVerbs.py

```diff
@@ -325,17 +325,284 @@ def listen(info):
+    elif {item1, item2} in {"canOpener", "tinCan"}:
+        item1 = "canOpener"
+        item2 = "tinCan"
+        itemData1 = []  # canOpener doesn't need data
+        itemData2 = getItemDataForUse(player, item2)
+    else:
+        itemData1 = []
+        itemData2 = []
+
+    # Look up item interaction response, if there is one
+    try:
+        response = itemData2[1].itemInteractions[itemData1[0]]
+        # If that doesn't get a valid response, try swapping the items
+        if response == "None":
+            response = itemData1[1].itemInteractions[itemData2[0]]
+    except (KeyError, ValueError, IndexError):
+        response = None
+
+    # Handle valid cases here
+    if response != "None":
+        if {item1, item2} == {"battery", "flashlight"}:
+            batteryFlashlight(player, game, itemData1, itemData2)
+        elif {item1, item2} == {"canOpener", "tinCan"}:
+            canOpenerCatFood(player, game, itemData2)
+        return response
+    else:
+        return None
+
+
+def isItem(itemName, allItems):
+    """
+    Helper function for 'useHandler' - returns boolean based on whether the
+    item is an "item" or a "feature".
+    """
+    if itemName in allItems:
+        return True
+    else:
+        return False
+
+
+def useHandler(player, game, allItems, item1, item2):
+    """
+    Helper function for 'use' - reads the type of items and processes
+    them depending on the type.
+    Returns a response based on the case.
+    """
+    type1 = isItem(item1, allItems)
+    type2 = isItem(item2, allItems)
+
+    # Prevent using something on itself
+    if item1 == item2:
+        response = "You can't use an item on itself."
+    # If one input is an Item, and the other a room feature
+    elif {type1, type2} == {True, False}:
+        response = combineItemAndFeature(player, item1, item2)
+    # If both item1 and item2 are Items
+    elif {type1, type2} == {True, True}:
+        response = combineTwoItems(player, game, item1, item2)
+    # Disallow using a feature on another feature
+    elif {type1, type2} == {False, False}:
+        response = "You can't use that."
+    else:
+        response = errorString
+
+    return response
+
+
+def getItemDataForUse(player, itemName):
+    """
+    Helper function for 'use' - gathers data about an item
+    that the player is trying to use.
+    Takes player and game data, and the item's name string.
+    Returns a list with all the item's necessary data.
+    """
+    inv = player.getInventory()
+    roomVis = player.getLocation().getVisibleItems()
+    roomDrop = player.getLocation().getDroppedItems()
+
+    # "itemUseData" list will hold, in this order:
+    # [0] itemName (str): the item name provided by player
+    # [1] itemObject (Item): the actual Item instance
+    # [2] location (str): "inv", "roomVis", or "roomDrop"
+    # [3] itemIndex (int): the current index of an item with multiple modes
+    itemUseData = [itemName]
```

Comment:
I chose the list instead because it specifically needs to hold an instance of Item, and as far as I'm aware, dicts only hold strings. It's a big part of why the function was separated into a helper function (to reduce complexity), so it also didn't make sense to translate strings to an Item instance on the receiving side.

I tried looking up how to have an instance of a class in a dict, but they just looked way more complicated than they needed to be for this purpose. So while I know it's not really fun to keep track of each index, it felt like the least awkward solution.

Any advice?

Reply:
Oh you can definitely store custom class objects as dictionary values. I'm pretty sure a Python dictionary can have any type object as a value. Here is an example I threw together really quickly:

![image](https://user-images.githubusercontent.com/61394328/202467224-9109b4c5-a613-4f03-8468-60d5cec0cb00.png)
![image](https://user-images.githubusercontent.com/61394328/202467347-f9989250-e51c-4e02-956e-f1580bb099df.png)

Reply:
This page has some more info about dictionaries in Python if you're curious. [https://realpython.com/python-dicts/#restrictions-on-dictionary-values](https://realpython.com/python-dicts/#restrictions-on-dictionary-values)

![image](https://user-images.githubusercontent.com/61394328/202468180-cb34b0ac-cdc1-42a9-b218-4c411e75dfb6.png)

Reply:
You're right, it works. Not sure where I went wrong before, but it's fixed. Thanks!